### PR TITLE
Checkstyle and github workflow action

### PR DESCRIPTION
**Pull request by RickPoleshuck**

### CODE REVIEW

**Comment by amfred** [comment]
There's a lot going on in this PR Rick, so it's hard to review - can you either break it up into a couple of PRs or summarize what's going on in the comments?

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+    build-prod:
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@v2
+            - name: Set up JDK 11
+              uses: actions/setup-java@v2
+              with:
+                  java-version: '11'
+                  distribution: 'adopt'
+            - run: mvn --batch-mode --update-snapshots -Pprod verify
```
Comment (amfred) [inline .github/workflows/build.yaml:13]: I'm new to Github Actions - does this actually deploy somewhere?

File: config/checkstyle.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+        <module name="ExplicitInitialization"/>
+        <module name="UnusedImports"/>
+        <module name="Regexp">
+            <property name="format" value="((System\.|(out|err)\.print)|printStackTrace)"/>
+            <property name="ignoreComments" value="true"/>
+            <property name="illegalPattern" value="true"/>
+            <property name="message" value="No printing to console, use a logger."/>
+        </module>
+    </module>
+</module>
```
Comment (amfred) [inline config/checkstyle.xml:21]: I'm not familiar with this particular linter, but it should be good to add.

File: config/checkstyle_suppressions.xml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+
+<!DOCTYPE suppressions PUBLIC
+        "-//Puppy Crawl//DTD Suppressions 1.1//EN"
+        "http://www.puppycrawl.com/dtds/suppressions_1_1.dtd">
+
+<suppressions>
+    <!-- io.swagger package is generated, don't run checks -->
+    <suppress files="[\\/]io[\\/]swagger[\\/]" checks="[a-zA-Z0-9]*"/>
+</suppressions>
```
Comment (amfred) [inline config/checkstyle_suppressions.xml:10]: Looks good to me.

File: src/main/java/application/model/AttorneyModel.java [hunk 4]
```diff
@@ -1,91 +1,92 @@
...
+    }
+
+    public QueryResult<Attorney> getAll() {
+        QueryResult<Attorney> qr = db.query(new QueryBuilder(ne("_id", "")).fields("username", "_id", "_rev").build(),
+                Attorney.class);
+        return qr;
+    }
+
+
 }
```
Comment (amfred) [inline src/main/java/application/model/AttorneyModel.java:92]: What's the intent of these changes?

File: src/main/java/application/model/ClientModel.java [hunk 5]
```diff
@@ -1,56 +1,55 @@
...
+    public Client read(String id) {
+        Client client = db.find(Client.class, id);
+        return client;
+    }
+
+    public QueryResult<Client> getAllClientsOfAttorney(String attorneyId) {
+        QueryResult<Client> qr = db.query(new QueryBuilder(eq("attorneyId", attorneyId)).build(), Client.class);
+        return qr;
+    }
 }
```
Comment (amfred) [inline src/main/java/application/model/ClientModel.java:55]: What's the intent of these changes?

File: src/main/java/application/rest/RootEndpoint.java [hunk 6]
```diff
@@ -13,11 +11,11 @@
...
-    return Response.ok("{\"health\":\"" + healthURL + "\",\"example\":\"" + exampleURL + "\"}").build();
-  }
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response listResources(@Context UriInfo uriInfo) {
+        String healthURL = (uriInfo.getAbsolutePath() + "/v1/health").replaceAll("(?<!http:)\\/\\/", "/");
+        String exampleURL = (uriInfo.getAbsolutePath() + "/v1/example").replaceAll("(?<!http:)\\/\\/", "/");
+        return Response.ok("{\"health\":\"" + healthURL + "\",\"example\":\"" + exampleURL + "\"}").build();
+    }
 }
```
Comment (amfred) [inline src/main/java/application/rest/RootEndpoint.java:21]: Is this just a simple formatting change then?

**Comment by RickPoleshuck** [comment]
> There's a lot going on in this PR Rick, so it's hard to review - can you summarize what's going on in the comments? I put my specific questions into my code review here.

There are 3 parts to this PR.
1. Checkstyle plugin and configuration was added to allow for static code quality analysis. As a first use of code quality analysis for this project, the checks are very limited. Look at checkstyle.xml if you want to see what particular checks are done.

2. The code was adjusted to pass all the checks implemented. No functionality was changed.

3. A github action was added to run these tests when pushed. This is performed by building the application, running all unit tests including code quality checks. No deployment is performed. Please look at https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html to understand the mvn lifecycle . 'verify' does not do an install or a deploy, but does perform an integration tests implemented.

**Comment by demilolu** [comment]
@kossiefu who should be the reviewer here from MS?

**Comment by kossiefu** [comment]
@RickPoleshuck, @upkarlidder do we still want to move forward with this change? It aligns with what the ms_cohort haas been working on. We can build on top of it.
